Action to connect to Twilio connection endpoint

andioneto opened the pull request:
### Actions Added
- **initiateTwilioCon:** 
-- params: userId (int), campaignId (int), volunteerCallActive (bool) [if true => throws error for already active connection)
-- **sends GET** req to 'users/${user_id}' and check to see if call_sid is set
-- if sid is set => thow error
-- if not set => next step
-- **sends POST** req to '/users/${userId}/campaigns/${campaignId}/calls'
-- if sid not set => throw error
-- if is set=> dispatches(setVolunteerActive)
- **endTwilioCon:** 
-- params: userId (int), campaignId (int), volunteerCallActive (bool) [if false => throws error for trying to shut an inactive connection) 
-- **sends GET** req to 'users/${user_id}' and check to see if call_sid is set
-- if sid is not set => throw error
-- if sid is set=> next step
-- **sends DELETE** req to '/users/${userId}/campaigns/${campaignId}/calls'
-- if sid is not set => dispatch(clearVolunteerActive)
-- if sid is set=> throw error

### Modifications to /public/src/components/sidebar_list_view/current_campaign:
- added in props for actions
- consolidated props being passed to current_campaign
- updated tests to reflect these changes
**TODO:** when endpoint for disconnect not yet finished, when it is, do walkthrough test

andioneto committed:
will hookup when 108 goes in

perfectlynormalbeast reviewed (changes requested):
We also need tests for this PR. Also the tests are failing for some reason?

andioneto commented:
Yeah, I think bc this is the really old version of the code I'm touching. I'm going to change this to WIP and either do it tonight or tomorrow AM. The actions can't dispatch the other actions they need (bc they don't yet exist in master), and I can't hook it up yet.

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-volunteercallaction

andioneto committed:
moew

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-volunteercallaction

andioneto committed:
updated current_campaign tests

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-volunteercallaction

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -102,6 +102,7 @@ export function releaseCall(userId, campaignId, callId, currentCallStatus, next
     if (next) {
       return dispatch(assignToCall(userId, campaignId));
     }
+    // dispatch(endVolunterTwilioCon(userId, campaignId, ))
```
Do we need this around?

andioneto replied:
This is what I'm adding.
So this is for when a user (instead of clicking skip) clicks 'stop calling' and wants to end their Twilio connection.

perfectlynormalbeast commented on public/src/actions/calls.jsx:
```diff
@@ -118,5 +119,27 @@ export function updateCallAttempt(userId, campaignId, callId, outcome, notes = n
+}
+
+export function initateVolunteerTwilioCon(userId, campaignId, volunteerCallActive) {
+  if (volunteerCallActive) {
```
So the issue here is that in some situations the client may end up with a stale idea of a volunteer's call state. It would be prudent to actually fetch a user's call state by GET on /users/:id and inspecting the call SID key in the response object.

Same comment for the method below. What do you think?

andioneto replied:
Cool beans, that sounds like a great idea

perfectlynormalbeast reviewed (changes requested):
One tiny comment and one larger concern

andioneto committed:
writing tests for new actions (connect to twilio)

andioneto committed:
working my way through updating after prs were merged
modified initiate and end volunteer connection actions to get volunteer's active status from server rather than relying on it from the store

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-volunteercallaction

andioneto committed:
rewrote flow for connecting/disconnecting to Twilio
added testing for checkTwilioCon, endTwilioCon, and initiateTwilioCon

perfectlynormalbeast reviewed (changes requested):
So I may have not been clear enough in my last review:
Having a method called `checkTwilioCon` also surreptitiously initiate/hang up calls based on 2 boolean params is a bad design pattern for 2 reasons: the name of the function, as well as the boolean flipping involved. It's a lot cleaner to have 2 methods (the ones you originally wrote) and them maybe internally calling a third function for checking the user's call SID and conditionally fail or proceed based on that. What do you think? This doesn't involve ripping out all the changes you made, just refactoring the methods involved and restoring the old action functions in the components.

andioneto committed:
refactored to remove the need for the check connection function
next: redo testing....

andioneto committed:
refactored to remove the need for the check connection function (#145)

next: redo testing....

andioneto committed:
fixed testing to reflect changes in calls actions

andioneto committed:
Merge branch 'ao-volunteercallaction' of https://github.com/Twilio-org/phonebank into ao-volunteercallaction

andioneto committed:
Merge branch 'master' of https://github.com/Twilio-org/phonebank into ao-volunteercallaction
